metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala

```diff
@@ -61,17 +63,28 @@ final class DefinitionProvider(
       token: CancelToken
   ): Future[DefinitionResult] = {
     val fromSemanticdb =
-      semanticdbs.textDocument(path).documentIncludingStale match {
-        case Some(doc) =>
-          definitionFromSnapshot(path, params, doc)
-        case _ =>
-          warnings.noSemanticdb(path)
-          DefinitionResult.empty
+      semanticdbs.textDocument(path).documentIncludingStale
+    val fromSnapshot = fromSemanticdb match {
+      case Some(doc) =>
+        definitionFromSnapshot(path, params, doc)
+      case _ =>
+        DefinitionResult.empty
+    }
+    val fromCompiler =
+      if (fromSnapshot.locations.isEmpty()) {
+        if (remote.isEnabledForPath(path)) {
+          remote.definition(params).map(_.getOrElse(fromSnapshot))
```

Comment:
I think we can do something like:

```scala
   val fromSemanticdb =
      semanticdbs.textDocument(path).documentIncludingStale
    val fromSnapshot = fromSemanticdb match {
      case Some(doc) =>
        definitionFromSnapshot(path, params, doc)
      case _ =>
        DefinitionResult.empty
    }
    val fromIndex =
      if (fromSnapshot.locations.isEmpty() && remote.isEnabledForPath(path)) {
          remote.definition(params).map(_.getOrElse(fromSnapshot))
      } else {
         fromSnapshot
        }
      } 
    if (fromIndex.locations.isEmpty()) {
      compilers().definition(params, token)
    } else {
      if(fromIndex.isEmpty)  warnings.noSemanticdb(path)
      Future.successful(fromIndex)
    }
```

This way it will also work for rambo mode, otherwise we assume that anything outside the build targets doesn't need any kind of compiler.

Reply:
This was a great suggestion! Applied (with minor changes)